test_conformance/subgroups/workgroup_common_templates.h

```diff
@@ -0,0 +1,778 @@
+            return a + b;
+        }
+        case ArithmeticOp::max_: {
+            return a > b ? a : b;
+        }
+        case ArithmeticOp::min_: {
+            return a < b ? a : b;
+        }
+        case ArithmeticOp::mul_: {
+            return a * b;
+        }
+        default: log_error("Unknown operation request"); break;
+    }
+    return 0;
+}
+
+template <>
+inline cl_float calculate(cl_float a, cl_float b, ArithmeticOp operation)
+{
+    switch (operation)
+    {
+        case ArithmeticOp::add_: {
+            return a + b;
+        }
+        case ArithmeticOp::max_: {
+            return a > b ? a : b;
+        }
+        case ArithmeticOp::min_: {
+            return a < b ? a : b;
+        }
+        case ArithmeticOp::mul_: {
+            return a * b;
+        }
+        default: log_error("Unknown operation request"); break;
+    }
+    return 0;
+}
+
+template <>
+inline subgroups::cl_half calculate(subgroups::cl_half a, subgroups::cl_half b,
+                                    ArithmeticOp operation)
+{
+    switch (operation)
+    {
+        case ArithmeticOp::add_: return to_half(to_float(a) + to_float(b));
+        case ArithmeticOp::max_:
+            return to_float(a) > to_float(b) || is_half_nan(b.data) ? a : b;
+        case ArithmeticOp::min_:
+            return to_float(a) < to_float(b) || is_half_nan(b.data) ? a : b;
+        case ArithmeticOp::mul_: return to_half(to_float(a) * to_float(b));
+        default: log_error("Unknown operation request"); break;
+    }
+    return to_half(0);
+}
+
+template <typename Ty> bool is_floating_point()
+{
+    return std::is_floating_point<Ty>::value
+        || std::is_same<Ty, subgroups::cl_half>::value;
+}
+
+template <typename Ty, ArithmeticOp operation>
+void genrand(Ty *x, Ty *t, cl_int *m, int ns, int nw, int ng)
+{
+    int nj = (nw + ns - 1) / ns;
+
+    for (int k = 0; k < ng; ++k)
+    {
+        for (int j = 0; j < nj; ++j)
+        {
+            int ii = j * ns;
+            int n = ii + ns > nw ? nw - ii : ns;
+
+            for (int i = 0; i < n; ++i)
+            {
+                cl_ulong x;
+                if ((operation == ArithmeticOp::add_
+                     || operation == ArithmeticOp::mul_)
+                    && is_floating_point<Ty>())
+                {
+                    // work around different results depending on operation
+                    // order by having input with little precision
+                    x = genrand_int32(gMTdata) % 64;
+                }
+                else
+                {
+                    x = genrand_int64(gMTdata);
+                    if ((operation == ArithmeticOp::logical_and
+                         || operation == ArithmeticOp::logical_or
+                         || operation == ArithmeticOp::logical_xor)
+                        && ((x >> 32) & 1) == 0)
+                        x = 0; // increase probability of false
+                }
+                set_value(t[ii + i], x);
```

Comment:
The data generated might be problematic in some cases. I specifically looked at some of the existing tests, such as for `sub_group_scan_(ex|in)clusive_(add|min|max)` and `sub_group_reduce_(add|min|max)`. There might be issues with other functions, too.

1. For the min/max functions with floating point types, the specification doesn't say much about NaN: i.e. are they propagated like with `fmax` or is the result undefined as with `max`? I suppose we should consider the second, weaker case given the lack of precision in the spec. With that in mind, the generated data should not included NaNs or we wouldn't be able to validate the output.
2. For the add functions on signed integer types, the generated values can be arbitrarily large and cause overflows and therefore undefined behaviour (on both the device and the host to compute the reference values). Previously, the generated values where small:
   - random values in `[1, subgroup_size]` for reduce or scan exclusive functions;
   - the sequence `0 1 2 ... (subgroup_size - 1)` for scan inclusive functions.

There might be other issues.

While the new random data is not perfect, it is arguably better than the previous one. It's also very nice to have this function instead of duplicated code here and there.

I'd argue this should be dealt with a follow-up task, but we should be aware of these issues, especially regarding undefined behaviour/results.